core/test/lib/libledger-test/CppHttpLibClient.cpp

```diff
@@ -0,0 +1,111 @@
+                }
+
+                std::unordered_map<std::string, std::string> getHeaders() override {
+                    return _headers;
+                }
+
+                core::api::HttpReadBodyResult readBody() override {
+                    std::vector<uint8_t> body((uint8_t *)_body.c_str(), (uint8_t *)(_body.c_str() + _body.size()));
+                    _body.clear();
+                    return core::api::HttpReadBodyResult(
+                    std::experimental::optional<core::api::Error>(),
+                    std::experimental::optional<std::vector<uint8_t>>(body)
+                    );
+                }
+
+            private:
+                int32_t _statusCode;
+                std::string _statusText;
+                std::unordered_map<std::string, std::string> _headers;
+                std::string _body;
+            };
+
+            void CppHttpLibClient::execute(const std::shared_ptr<core::api::HttpRequest> &request) {
+            _context->execute(make_runnable([=] () {
+                auto url = request->getUrl();
+                std::size_t sep;
+                std::size_t sepIgnored = url.find("//");
+                if (sepIgnored == std::string::npos) {
+                    sep = url.find('/');
+                }
+                else {
+                    sep = url.find('/', sepIgnored+2);
+                }
+                auto baseUrl = url.substr(0, sep);
+                auto path = url.substr(sep);
+                
+                httplib::Client cli(baseUrl.c_str());
+                httplib::Headers headers;
+                for (const auto& header : request->getHeaders()) {
+                    headers.insert(std::make_pair(header.first, header.second));
+                }
+                cli.set_default_headers(headers);
+                std::string body;
+                if (request->getBody().size() > 0) {
+                    std::stringstream bodyStream;
+                    for (auto byte : request->getBody()) {
+                        bodyStream << (char )byte;
+                    }
+                    body = bodyStream.str();
+                }
+                
+                httplib::Result res = 
+                    (request->getMethod() == core::api::HttpMethod::POST) ? cli.Post(path.c_str(), body, "text/plain") :
+                    (request->getMethod() == core::api::HttpMethod::GET) ?  cli.Get(path.c_str()) :
+                    (request->getMethod() == core::api::HttpMethod::PUT) ?  cli.Put(path.c_str(), body, "text/plain") :
+                                                                            cli.Delete(path.c_str());
+                
+                if (res.error() != httplib::Error::Success || res->status < 200 || res->status >= 300 ) {
+                    request->complete(nullptr, std::experimental::optional<core::api::Error>(
+                        core::api::Error(core::api::ErrorCode::HTTP_ERROR, res->reason)));       

```

Comment:
For the second param here, you should build a string containing both the status and reason, something like:

`fmt::format("{}: {}", res->status, res->reason)`

Both will be useful info in case of failure.